### PR TITLE
Experimental support Rx

### LINKED ISSUE #114: Implement support for IObservable<T>
The current design has a lot of different APIs for subscribing to event streams, filtering them in various ways then performing actions. This needs a lot of documentation, implementation, test, etc and makes things more complex for people trying to figure out how to do stuff.

Reactive Extensions is a standard .NET library for managing asynchronous event-based programming. By supporting this natively netdaemon could leverage the existing library and documentation and not have to reimplement the basic functionality. More info can be found at https://github.com/dotnet/reactive, http://introtorx.com/ and http://reactivex.io/

## Additional context
The current Fluent API IStateChanged interface has multiple overloads of WhenStateChange.
```
Entity("binary_sensor.my_motion_sensor").WhenStateChange(to: "on").UseEntity("light.light1").TurnOn().Execute();
Entity("binary_sensor.my_motion_sensor").WhenStateChange(from: "off", to:"on").UseEntity("light.light1").TurnOn().Execute();
Entity("binary_sensor.my_motion_sensor").WhenStateChange((toState, fromState) => toState?.State == "on").UseEntity("light.light1").TurnOn().Execute();
Entity("sun.sun")
    .WhenStateChange((n, o) =>
        n?.Attribute?.elevation <= 3.0 &&
        n?.Attribute?.rising == false &&
        o?.Attribute?.elevation >  3.0).UseEntity("light.light1").TurnOn().Execute();
```
Using RX these can all just be replaced by a standard LINQ Where() filter. For example, if IEntity implemented `IObservable<EntityState>` then you can do something like this
```
Entity("binary_sensor.my_motion_sensor")
    .Where(state => state.State == "on")
    .ForEach(e => Entity("light.light1").TurnOn());
```
For the case where your logic needs the previous state, the current API has a WhenStateChange overload that provides it as toState and fromState. With RX that is not required as you can use the standard RX APIs to create whatever kind of event stream you need:
```
Entity("binary_sensor.my_motion_sensor")
    .Scan((default(EntityState),default(EntityState)), (last, current) => (last.Item2, current))
    .Where(t => t.Item2.State == "on")
    .ForEach(e => Entity("light.light1").TurnOn());
```
Here the Scan() operator is turning our IObservable<EntityState> event stream into a `IObservable<(EntityState,EntityState)>` event stream, where each event is a tuple (from, to). There is no special code required in netdaemon to handle the previous state (although you could provide some convenience wrapper if it was a common usage). If I wanted a rolling average or the last 5 measurements or debouncing, I can build that directly.

Let's consider a more complicated (but realistic!) example. I have a light that I want on from sunset until 10pm, but only when I am home.
```
var sunset = Entity("sun.sun").Where(s => s.State == "Below horizon").Select(s => true);
var tenpm = new DateTime(DateTime.Today.Year, DateTime.Today.Month, DateTime.Today.Day, 22, 0, 0);
var bedtime = Observable.Timer(new DateTimeOffset(tenpm), TimeSpan.FromDays(1)).Select(s => false);
var imhome = Entity("person.john").Where(s => s.State == "home").Select(s => true);
```
Now we have created three different IObservable<bool> event streams. sunset emits TRUE at sunset. bedtime emits FALSE at 10pm. And imhome emits TRUE or FALSE depending on my presence. Now we just have to combine them to get the state we want the light to be in.
```
sunset.Concat(bedtime)
    .CombineLatest(imhome, (time, home) => time && home)
    .ForEach(turnon => {
        if (b) {
            Entity("light.light1").TurnOn());
        } else {
            Entity("light.light1").TurnOff());
        }
    };
```
Again you can see that once these event streams exist they can easily be combined in interesting ways without having to add any additional code to netdaemon. There might be some convenient wrappers for common patterns, but the underlying plumbing is all done by RX.

### PR DESCRIPTION
**Pull request by helto4real**
## Proposed change

** I know this is way too big PR to be a good one... **

## Refactorings of where logic is put
In order to make application lifecycle easier, all eventlisteners and state handlers are tied to each app-instance. This makes it way more easy to disable individual apps in runtime plus manage lifecykle in the Rx stuff too. 

## Implement the Rx API
Implement the Rx interface. This is an alternative implementation to @jvert suggestions (thanks alot for this great idéa). This proposal was so different it was easier to implement it in own PR rather than change yours.. Hope that is ok :). 

New BaseClass `NetDaemonRxApp` that will implement the Rx style. We do not want to mix with old style. This is event based and not Async/await. All calls post a message on a Channel and async code takes care of the async stuff. 

Tests will be added. after feedback.
More API features will be added before merge. 

Example of usage:

```c#
// Observable that get state changes 
StateChanges.Subscribe(t=> Log(t.New.State));

// Observable that get all state changes inkluding attributes 
StateAllChanges.Subscribe(t=> Log(t.New.State));

// IEnumerable<EntityState>
var allLights = States.Select(n => n.EntityId.StartsWith("light."));
// Gets single state
var state = State("light.my_light")?.State;

// No async, handled in background
CallService("light", "turn_on", new {entity_id = "light.my_light"});

// Entity now not fluent
// Action on single entity
Entity("light.my_light").TurnOn();
Entity("light.my_light").Toggle();
// Action on multiple entities
Entities("light.my_light", "light.my_light").Toggle();
// Or lambda
Entities(n => n.EntityId.StartsWith("light.").Toggle();

// Merging observables <3
Entities(
    "binary_sensor.tomas_rum_pir",
    "binary_sensor.vardagsrum_pir")
    .StateChanges
    .Subscribe(e =>
    {
        Log("{entity}: {state}({oldstate}, {lastchanged})", e.New.EntityId, e.New.State, e.Old.State, e.New.LastChanged);
    });

// Merging observables all changes including attributes <3
Entities(
    "binary_sensor.tomas_rum_pir",
    "binary_sensor.vardagsrum_pir")
    .StateAllChanges
    .Subscribe(e =>
    {
        Log("{entity}: {state}({oldstate}, {lastchanged})", e.New.EntityId, e.New.State, e.Old.State, e.New.LastChanged);
    });

// Set state
SetState("sensor.thesensor", "on", new {attributex="cool"});
// Set state selecting with Entity Selector
Entity("sensor.x", "sensor.y").SetState("on");

// Merging of entity results

// Schedulers
RunEvery(TimeSpan.FromMinutes(1)).Subscribe(....);

RunDaily("12:00:00").Subscribe(...);

// Events
EventChanges
     .Subscribe(f =>
      {
            Log("event: {domain}.{event} - {data}", f?.Domain??"none", f.Event, f?.Data);
       });


```

## Type of change

- [X] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (which adds functionality to an existing integration)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #114 
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code compiles without warnings (code quality chek)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/net-daemon/docs

### CODE REVIEW

**Comment by danpowell88** [comment]
It's looking fairly nice and clean to me

**Comment by helto4real** [comment]
> It's looking fairly nice and clean to me

Thank you for taking time to comment.

File: src/App/NetDaemon.App/Common/Reactive/INetDaemonReactive.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    ///     Interface for Observable Events
+    /// </summary>
+    public interface IRxEvent : IObservable<RxEvent>
+    {
+    }
+
+    /// <summary>
+    ///     Interface for scheduling
+    /// </summary>
+    public interface IRxSchedule
```
Comment (jvert) [inline src/App/NetDaemon.App/Common/Reactive/INetDaemonReactive.cs:107]: These are super useful! It doesn't say what the 'long' represents though - I assume a Windows file time? Might be a little more obvious if it was IObservable<DateTime> but I'm not sure how useful that is - you could just do DateTime.Now.

File: src/App/NetDaemon.App/Common/Reactive/RxEntity.cs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        /// <summary>
+        ///     Observable, All state changes. New.State!=Old.State
+        /// </summary>
+        IObservable<(EntityState Old, EntityState New)> StateChanges { get; }
+    }
+
+    /// <summary>
+    ///     Interface for objects implements SetState
+    /// </summary>
+    public interface ISetState
```
Comment (jvert) [inline src/App/NetDaemon.App/Common/Reactive/RxEntity.cs:27]: Is there a reason for so many different interfaces? Why not 
```
public interface ISetState {
    void SetState(dynamic state, dynamic? attributes);
    void Toggle(dynamic? attributes);
    void TurnOff(dynamic? attributes);
    void TurnOn(dynamic? attributes);
}
```
or is there really a scenario where something would implement ITurnOn but not ITurnOff?

File: src/App/NetDaemon.App/Common/Reactive/RxEntity.cs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+                throw new Exception("StateAllChanges not allowed. No entity id:s were selected. Check your lambda query");
+            }
+        }
+
+        /// <inheritdoc/>
+        public IObservable<(EntityState Old, EntityState New)> StateChanges
+        {
+            get
+            {
+                if (_entityIds.Count() > 1)
```
Comment (jvert) [inline src/App/NetDaemon.App/Common/Reactive/RxEntity.cs:122]: you can get rid of all the resultList/Observable.Merge stuff and just do
```
return _daemon.StateChanges.Where(f => _entityIds.Contains(f.New.EntityId) && f.New.State != f.Old.State)
```

File: src/App/NetDaemon.App/Common/NetDaemonApp.cs [hunk 4]
```diff
@@ -16,67 +16,21 @@ namespace JoySoftware.HomeAssistant.NetDaemon.Common
...
-        /// <inheritdoc/>
-        public dynamic Storage => _storageObject ?? throw new NullReferenceException($"{nameof(_storageObject)} cant be null");
-
-        internal Channel<bool> InternalLazyStoreStateQueue => _lazyStoreStateQueue;
-        internal FluentExpandoObject? InternalStorageObject { get { return _storageObject; } set { _storageObject = value; } }
-
+ 
         /// <inheritdoc/>
-        public Task CallService(string domain, string service, dynamic? data = null, bool waitForResponse = false)
+        public Task CallServiceAsync(string domain, string service, dynamic? data = null, bool waitForResponse = false)
```
[the remark this answers is not in the document]
Comment (helto4real) [reply src/App/NetDaemon.App/Common/NetDaemonApp.cs:30]: This is the old Api that I am changing the names to standard Async postfix. Most functions that are async use that postfix. I want to make the very different from the new API

File: src/App/NetDaemon.App/Common/Reactive/INetDaemonReactive.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        ///     Entity to perform actions on
+        /// </summary>
+        /// <param name="entityId">EntityId</param>
+        RxEntity Entity(string entityId);
+    }
+
+    /// <summary>
+    ///     Interface for Observable Events
+    /// </summary>
+    public interface IRxEvent : IObservable<RxEvent>
```
[the remark this answers is not in the document]
Comment (helto4real) [reply src/App/NetDaemon.App/Common/Reactive/INetDaemonReactive.cs:112]: Hmm there was a reason I can not remember now... I will check that out

File: src/App/NetDaemon.App/Common/Reactive/RxEntity.cs (hunk 2 above)
Comment (helto4real) [reply src/App/NetDaemon.App/Common/Reactive/RxEntity.cs:27]: I am thinking I will reuse them maybe on the other API.. If not I will consolidate..

File: src/App/NetDaemon.App/Common/Reactive/RxEntity.cs (hunk 3 above)
Comment (helto4real) [reply src/App/NetDaemon.App/Common/Reactive/RxEntity.cs:122]: Allright... that is a good suggestion

**Pull request merged**